src/VectorBundle.jl

```diff
@@ -0,0 +1,372 @@
+* `copyto!(v, w)` for vectors `v` and `w`,
+* `eltype(v)` for vector `v`,
+* `vector_space_dimension(::VectorBundleFibers{<:typeof(VS)})`.
+
+Optionally:
+* inner product via `inner` (used to provide Riemannian metric on vector
+  bundles),
+* flat_isomorphism! and sharp_isomorphism!,
+* norm (by default uses `inner`),
+* project_vector! (for embedded vector spaces),
+* representation_size (if support for `ProductArray` is desired),
+* broadcasting for basic operations.
+"""
+abstract type VectorSpaceType end
+
+struct TangentSpaceType <: VectorSpaceType end
+struct CotangentSpaceType <: VectorSpaceType end
+
+TCoTSpaceType = Union{TangentSpaceType, CotangentSpaceType}
+
+const TangentSpace = TangentSpaceType()
+const CotangentSpace = CotangentSpaceType()
+
+"""
+    TensorProductType(spaces::VectorSpaceType...)
+
+Vector space type corresponding to the tensor product of given vector space
+types.
+"""
+struct TensorProductType{TS<:Tuple} <: VectorSpaceType
+    spaces::TS
+end
+
+TensorProductType(spaces::VectorSpaceType...) = TensorProductType{typeof(spaces)}(spaces)
+
+"""
+    VectorBundleFibers(VS::VectorSpaceType, M::Manifold)
+
+Type representing a family of vector spaces (fibers) of a vector bundle over `M`
+with vector spaces of type `VS`. In contrast with `VectorBundle`, operations
+on `VectorBundleFibers` expect point-like and vector-like parts to be
+passed separately instead of being bundled together. It can be thought of
+as a representation of vector spaces from a vector bundle but without
+storing the point at which a vector space is attached (which is specified
+separately in various functions).
+"""
+struct VectorBundleFibers{TVS<:VectorSpaceType, TM<:Manifold}
+    VS::TVS
+    M::TM
+end
+
+"""
+    VectorSpaceAtPoint(fiber::VectorBundleFibers, x)
+
+A vector space (fiber type `fiber` of a vector bundle) at point `x` from
+the manifold `fiber.M`.
+"""
+struct VectorSpaceAtPoint{TFiber<:VectorBundleFibers, TX}
+    fiber::TFiber
+    x::TX
+end
+
+"""
+    FVector(type::VectorSpaceType, data)
+
+Decorator indicating that the vector `data` from a fiber of a vector bundle
+is of type `type`.
+"""
+struct FVector{TType<:VectorSpaceType,TData}
+    type::TType
+    data::TData
+end
+
+size(x::FVector) = size(x.data)
+Base.@propagate_inbounds getindex(x::FVector, i) = getindex(x.data, i)
+Base.@propagate_inbounds setindex!(x::FVector, val, i) = setindex!(x.data, val, i)
+
+(+)(v1::FVector, v2::FVector) = FVector(v1.type, v1.data + v2.data)
+(-)(v1::FVector, v2::FVector) = FVector(v1.type, v1.data - v2.data)
+(-)(v::FVector) = FVector(v.type, -v.data)
+(*)(a::Number, v::FVector) = FVector(v.type, a*v.data)
+
+eltype(::Type{FVector{TType,TData}}) where {TType<:VectorSpaceType,TData} = eltype(TData)
+similar(x::FVector) = FVector(x.type, similar(x.data))
+similar(x::FVector, ::Type{T}) where {T} = FVector(x.type, similar(x.data, T))
+
+@doc doc"""
+    flat_isomorphism!(M::Manifold, v::FVector, x, w::FVector)
```

Comment:
Since isomorphism is a property of the map that won't always be included in the name of isomorphisms, wouldn't it make more sense to simply name these functions `flat` and `sharp`? Also, I think it would be nice to have more documentation here. sharp and flat are defined such that they produce (co)vectors such that the usual covector action upon a vector is equivalent to the inner product (but with equations :)).

Reply:
I would also op for `flat` and `sharp` – while sharp could even be `♭`, for sharp, that's not so easy ;)

Reply:
Eh, it's not so hard. `\sharp` is `♯`, not `#`, so it works.

Reply:
Right, I forgot. And I wanted to start calling `#` octothorpe ( https://en.wikipedia.org/wiki/Number_sign )

Reply:
Can we all please agree to call it _anything_ before "octothorpe" ;)

Reply:
Yes, of course, that's for `#` not for `♯`.